### PR TITLE
Aegis 6406 asm2vec pytorch retrain

### PR DESCRIPTION
**Pull request by ilektragiassa**
Create scripts for:
1. training an asm2vec model
2. calculating the tensor representations

### CODE REVIEW

File: asm2vec/tensors.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import os
+import torch
+import logging
+import asm2vec
```
Comment (jamienutter) [inline asm2vec/tensors.py]: You shouldn't have this open import for the package your working inside... will create circular dependencies

File: asm2vec/tensors.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    :param epochs: number of epochs
+    :param device:  'auto' | 'cuda' | 'cpu'
+    :param lr: learning rate
+    """
+    tensors_list = []
+    if device == 'auto':
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    if os.path.isfile(model_path):
+        model, tokens = asm2vec.utils.load_model(model_path, device=device)
```
Comment (jamienutter) [inline asm2vec/tensors.py]: Just use the utils you imported
```suggestion
        model, tokens = utils.load_model(model_path, device=device)
```

File: asm2vec/tensors.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import torch
+import logging
+import asm2vec
+from asm2vec import utils
+from pathlib import Path
+
+logging.basicConfig(level=logging.INFO, format='%(message)s')
+
+
+def calc_tensors(asm_path, tensor_path, model_path, epochs, device='cpu', lr=0.02) -> list:
```
Comment (jamienutter) [inline asm2vec/tensors.py]: Can you include types for your args?
```suggestion
def calc_tensors(asm_path: str, tensor_path: str, model_path: str, epochs: int, device: str = 'cpu', learning_rate: float = 0.02) -> list:
```
And `lr` is not a good arg name, too unclear. Can we update to something like `learning_rate`? Verbose is better with naming

File: asm2vec/train.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import torch
+import asm2vec
```
Comment (jamienutter) [inline asm2vec/train.py]: Don't want this import

File: asm2vec/train.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    :param device: 'auto' | 'cuda' | 'cpu'
+    :param lr: learning rate
+    """
+
+    if device == 'auto':
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    if model_path:
+        model, tokens = asm2vec.utils.load_model(model_path, device=device)
+        functions, tokens_new = asm2vec.utils.load_data(train_set, limit=limit)
```
Comment (jamienutter) [inline asm2vec/train.py]:
```suggestion
        model, tokens = utils.load_model(model_path, device=device)
        functions, tokens_new = utils.load_data(train_set, limit=limit)
```

File: asm2vec/train.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        model=model,
+        embedding_size=embedding_size,
+        batch_size=batch_size,
+        epochs=epochs,
+        neg_sample_num=neg_sample,
+        calc_acc=calc_acc,
+        device=device,
+        callback=callback,
+        learning_rate=lr
+    )
```
Comment (jamienutter) [inline asm2vec/train.py]: If making a `train` module then do you not want to move the `utils.train` method into this module? 
I would

File: asm2vec/train.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    if model_path:
+        model, tokens = asm2vec.utils.load_model(model_path, device=device)
+        functions, tokens_new = asm2vec.utils.load_data(train_set, limit=limit)
+        tokens.update(tokens_new)
+        model.update(len(functions), tokens.size())
+    else:
+        model = None
+        functions, tokens = asm2vec.utils.load_data(Path(train_set), limit=limit)
```
Comment (jamienutter) [inline asm2vec/train.py]:
```suggestion
        functions, tokens = utils.load_data(Path(train_set), limit=limit)
```

File: asm2vec/train.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        embedding_size=embedding_size,
+        batch_size=batch_size,
+        epochs=epochs,
+        neg_sample_num=neg_sample,
+        calc_acc=calc_acc,
+        device=device,
+        callback=callback,
+        learning_rate=lr
+    )
+    asm2vec.utils.save_model(new_model, model, tokens)
```
Comment (jamienutter) [inline asm2vec/train.py]:
```suggestion
   utils.save_model(new_model, model, tokens)
```

File: asm2vec/train.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        model_path: str | None,
+        limit: int,
+        epochs: int,
+        calc_acc: False,
+        embedding_size=100,
+        batch_size=1024,
+        neg_sample=25,
+        lr=0.02,
+        device='cpu',
+) -> None:
```
Comment (jamienutter) [inline asm2vec/train.py]: Why do we not return the trained model?

File: asm2vec/train.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        new_model: str,
+        model_path: str | None,
+        limit: int,
+        epochs: int,
+        calc_acc: False,
+        embedding_size=100,
+        batch_size=1024,
+        neg_sample=25,
+        lr=0.02,
+        device='cpu',
```
Comment (jamienutter) [inline asm2vec/train.py]: types?

and `lr` -> `learning_rate`

File: asm2vec/tensors.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    if not (os.path.exists(dir0)):
+        os.mkdir(dir0)
+
+    if os.path.isdir(asm_path):
+        obj = os.scandir(asm_path)
+        for entry in obj:
+            if entry.is_dir() and os.listdir(entry) and entry.name:
+                tensor_file = os.path.join(dir0, entry.name)
+                if not (os.path.exists(tensor_file)):
+                    functions, tokens_new = asm2vec.utils.load_data([entry])
```
Comment (jamienutter) [inline asm2vec/tensors.py]:
```suggestion
                    functions, tokens_new = utils.load_data([entry])
```

**Review by jamienutter — CHANGES_REQUESTED** [review]

File: asm2vec/train.py (hunk 6 above)
Comment (ilektragiassa) [reply asm2vec/train.py]: `utils` is also used by the scripts remaining in the scripts directory, thus it will be complicated if it is moved

Comment (jamienutter) [reply asm2vec/train.py]: You can just change the import. 
`from asm2vec.train import train`

Comment (ilektragiassa) [reply asm2vec/train.py]: Will fix the contents of the `scripts` folder in the next PR

**Review by jamienutter — APPROVED** [review]

**Pull request merged**